**Question 3**

Let's use the model!

Write a script for loading the pipeline with pickle

Score this record:
{
    "lead_source": "paid_ads",
    "number_of_courses_viewed": 2,
    "annual_income": 79276.0
}

What's the probability that this lead will convert?

- 0.333
- **0.533**
- 0.733
- 0.933

If you're getting errors when unpickling the files, check their checksum:

$ md5sum pipeline_v1.bin
7d17d2e4dfbaf1e408e1a62e6e880d49 *pipeline_v1.bin

In [ ]:
import pickle

# 1. Load the pipeline
with open('pipeline_v1.bin', 'rb') as f:
    model = pickle.load(f)

# 2. Input record
record = {
    "lead_source": "paid_ads",
    "number_of_courses_viewed": 2,
    "annual_income": 79276.0
}

# 3. Convert record to list of dicts and predict
X = [record]
y_pred = model.predict_proba(X)[0, 1]

print("Probability that this lead will convert:", y_pred)

Probability that this lead will convert: 0.5336072702798061


## Question 4

Now let's serve this model as a web service

* Install FastAPI
* Write FastAPI code for serving the model
* Now score this client using `requests`:

```python
url = "YOUR_URL"
client = {
    "lead_source": "organic_search",
    "number_of_courses_viewed": 4,
    "annual_income": 80304.0
}
requests.post(url, json=client).json()
```

What's the probability that this client will get a subscription?

* 0.334
* **0.534**
* 0.734
* 0.934


## Question 5

Download the base image `agrigorev/zoomcamp-model:2025`. You can easily make it by using [docker pull](https://docs.docker.com/engine/reference/commandline/pull/) command.

So what's the size of this base image?

* 45 MB
* **121 MB**
* 245 MB
* 330 MB

You can get this information when running `docker images` - it'll be in the "SIZE" column.

## Dockerfile

Now create your own `Dockerfile` based on the image we prepared.

It should start like that:

```docker
FROM agrigorev/zoomcamp-model:2025
# add your stuff here
```

Now complete it:

* Install all the dependencies from pyproject.toml
* Copy your FastAPI script
* Run it with uvicorn 

After that, you can build your docker image.


## Question 6

Let's run your docker container!

After running it, score this client once again:

```python
url = "YOUR_URL"
client = {
    "lead_source": "organic_search",
    "number_of_courses_viewed": 4,
    "annual_income": 80304.0
}
requests.post(url, json=client).json()
```

What's the probability that this lead will convert?

* 0.39
* **0.59**
* 0.79
* 0.99


In [3]:
import requests
url = "http://0.0.0.0:9696/predict"
client = {
    "lead_source": "organic_search",
    "number_of_courses_viewed": 4,
    "annual_income": 80304.0
}
response = requests.post(url, json=client).json()

print(response)

if response['subscription_probability'] >= 0.5:
    print('customer is likely to subscribe, send promo')
    print(response['subscription_probability'])
else:
    print('customer is not likely to subscribe')
    print(response['subscription_probability'])


{'subscription_probability': 0.5340417283801275}
customer is likely to subscribe, send promo
0.5340417283801275
